File: csharp/OrderDispatchKata.Tests/UseCase/OrderShipmentUseCaseTest.cs
using NUnit.Framework;
using OrderDispatchKata.Domain;
using OrderDispatchKata.Tests.Doubles;
using OrderDispatchKata.UseCase;

namespace OrderDispatchKata.Tests.UseCase;

[TestFixture]
public class OrderShipmentUseCaseTest
{
    [SetUp]
    public void SetUp()
    {
        orderRepository = new TestOrderRepository();
        shipmentService = new TestShipmentService();
        useCase = new OrderShipmentUseCase(orderRepository, shipmentService);
    }

    private TestOrderRepository orderRepository;
    private TestShipmentService shipmentService;
    private OrderShipmentUseCase useCase;

    [Test]
    public void shipApprovedOrder()
    {
        var initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.APPROVED);
        orderRepository.addOrder(initialOrder);

        var request = new OrderShipmentRequest();
        request.setOrderId(1);

        useCase.run(request);

        Assert.That(orderRepository.getSavedOrder().getStatus(), Is.EqualTo(OrderStatus.SHIPPED));
        Assert.That(shipmentService.getShippedOrder(), Is.EqualTo(initialOrder));
    }

    [Test]
    public void createdOrdersCannotBeShipped()
    {
        var initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.CREATED);
        orderRepository.addOrder(initialOrder);

        var request = new OrderShipmentRequest();
        request.setOrderId(1);

        Assert.That(() => useCase.run(request),
            Throws.TypeOf<OrderCannotBeShippedException>());


        Assert.That(orderRepository.getSavedOrder(), Is.Null);
        Assert.That(shipmentService.getShippedOrder(), Is.Null);
    }

    [Test]
    public void rejectedOrdersCannotBeShipped()
    {
        var initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.REJECTED);
        orderRepository.addOrder(initialOrder);

        var request = new OrderShipmentRequest();
        request.setOrderId(1);

        Assert.That(() => useCase.run(request),
            Throws.TypeOf<OrderCannotBeShippedException>());
    }

    [Test]
    public void shippedOrdersCannotBeShippedAgain()
    {
        var initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.SHIPPED);
        orderRepository.addOrder(initialOrder);

        var request = new OrderShipmentRequest();
        request.setOrderId(1);

        Assert.That(() => useCase.run(request),
            Throws.TypeOf<OrderCannotBeShippedTwiceException>());
    }
}

File: java-lombok/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCaseTest.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository;
import it.gabrieletondi.telldontaskkata.doubles.TestShipmentService;
import org.junit.Test;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;

public class OrderShipmentUseCaseTest {
    private final TestOrderRepository orderRepository = new TestOrderRepository();
    private final TestShipmentService shipmentService = new TestShipmentService();
    private final OrderShipmentUseCase useCase = new OrderShipmentUseCase(orderRepository, shipmentService);

    @Test
    public void shipsApprovedOrder() {
        Order initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.APPROVED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder().getStatus(), is(OrderStatus.SHIPPED));
        assertThat(shipmentService.getShippedOrder(), is(initialOrder));
    }

    @Test(expected = OrderNotShippable.class)
    public void createdOrdersCannotBeShipped() {
        Order initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.CREATED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }

    @Test(expected = OrderNotShippable.class)
    public void rejectedOrdersCannotBeShipped() {
        Order initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.REJECTED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }

    @Test(expected = OrderAlreadyShipped.class)
    public void shippedOrdersCannotBeShippedAgain() {
        Order initialOrder = new Order();
        initialOrder.setId(1);
        initialOrder.setStatus(OrderStatus.SHIPPED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }
}
